File: python/0046.py
class Solution(object):
    def __init__(self):
        self.ans=[]
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def go(nums,beg):
            if beg==len(nums):
                self.ans.append(nums[:])
                return
            for i in range(beg,len(nums)):
                nums[beg],nums[i]=nums[i],nums[beg]
                go(nums,beg+1)
                nums[beg],nums[i]=nums[i],nums[beg]
        go(nums,0)
        return self.ans

File: rust/src/bin/0046.rs
struct Solution;

fn go(ans: &mut Vec<Vec<i32>>, ind: usize, a: &mut Vec<i32>) {
    if ind == a.len() {
        ans.push(a.clone());
        return;
    }
    for i in ind..a.len() {
        a.swap(ind, i);
        go(ans, ind + 1, a);
        a.swap(ind, i);
    }
}

impl Solution {
    pub fn permute(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans: Vec<Vec<i32>> = vec![];
        go(&mut ans, 0, &mut nums);
        ans
    }
}

fn main() {
    let ans = Solution::permute(vec!(1, 2, 3));
    println!("{:?}", ans);
}